jme3-examples/src/main/java/jme3test/terrain/PBRTerrainAdvancedTest.java

```diff
@@ -0,0 +1,429 @@
+import com.jme3.input.controls.KeyTrigger;
+import com.jme3.light.AmbientLight;
+import com.jme3.light.DirectionalLight;
+import com.jme3.light.LightProbe;
+import com.jme3.light.LightProbe.AreaType;
+import com.jme3.material.Material;
+import com.jme3.math.ColorRGBA;
+import com.jme3.math.Vector3f;
+import com.jme3.scene.Node;
+import com.jme3.shader.VarType;
+import com.jme3.system.AppSettings;
+import com.jme3.terrain.geomipmap.TerrainQuad;
+import com.jme3.terrain.heightmap.AbstractHeightMap;
+import com.jme3.terrain.heightmap.ImageBasedHeightMap;
+import com.jme3.texture.Image;
+import com.jme3.texture.Texture;
+import com.jme3.texture.Texture.WrapMode;
+import com.jme3.texture.TextureArray;
+import java.util.ArrayList;
+import java.util.List;
+
+
+// Uses assets from CC0Textures.com, licensed under CC0 1.0 Universal.
+
+public class PBRTerrainAdvancedTest extends SimpleApplication {    
+    
+    
+    //similar to PRBTerrainTest, but this one uses texture arrays to allow for more texture slots. beware that GPUs for older devices and mobile devices may be less likely to support texture arrays
+        
+    //results look better with anti-aliasing, especially at far distances. 
+    
+    private TerrainQuad terrain;
+    private Material matTerrain;
+    private boolean triPlanar = false;
+    private float dirtScale = 24;
+    private float darkRockScale = 24;
+    private float snowScale = 64;
+    private float tileRoadScale = 64;
+    private float grassScale = 24;
+    private float marbleScale = 64;
+    private float gravelScale = 64;
+    
+    private AmbientLight ambientLight;
+    private DirectionalLight directionalLight;
+    private boolean isNight = false;
+    private final float dayLightIntensity = 1.0f;
+    private final float nightLightIntensity = 0.03f;
+    
+    private BitmapText keybindingsText;
+
+    public static void main(String[] args) {
+        PBRTerrainAdvancedTest app = new PBRTerrainAdvancedTest();
+        AppSettings s = new AppSettings(true);
+        
+        s.put("GammaCorrection", true);
+        
+        app.setSettings(s);
+        app.start();
+        
+ 
+       
+    }
+    
+    private final ActionListener actionListener = new ActionListener() {
+        @Override
+        public void onAction(String name, boolean pressed, float tpf) {
+            if (name.equals("triPlanar") && !pressed) {
+                triPlanar = !triPlanar;
+                if (triPlanar) {
+                    matTerrain.setBoolean("useTriPlanarMapping", true);
+                    // planar textures don't use the mesh's texture coordinates but real world coordinates,
+                    // so we need to convert these texture coordinate scales into real world scales so it looks
+                    // the same when we switch to/from tr-planar mode (1024f is the alphamap size)
+                    matTerrain.setFloat("AlbedoMap_0_scale", 1f /  (1024f / dirtScale));
```

Comment:
The expression can be simplified to `dirtScale / 1024f'. The following statements can be similarly simplified.